File: apps/cargo-tracker/src/main/java/com/example/cargotracker/shipper/domain/model/Shipper.java
package com.example.cargotracker.shipper.domain.model;

import com.example.cargotracker.shared.domain.model.ShipperId;
import javax.annotation.concurrent.Immutable;

/**
 * 荷主。Shipper Context の集約ルート。
 *
 * <p>他の境界付けられたコンテキストのクラスを直接参照しない。Booking からの参照は
 * {@code ShipperExistenceChecker} ACL ポートを経由する（{@code domain-model.md}）。
 *
 * <p>法人荷主（契約番号・契約割引率）は US03 で扱う。本クラスは種別を保持するのみとする。
 */
@Immutable
public final class Shipper {

    private final ShipperId id;
    private final ShipperCode shipperCode;
    private final ShipperType shipperType;
    private final ShipperName name;
    private final Email email;
    private final Phone phone;
    private final Address address;

    public Shipper(
            ShipperId id,
            ShipperCode shipperCode,
            ShipperType shipperType,
            ShipperName name,
            Email email,
            Phone phone,
            Address address) {
        if (id == null || shipperCode == null || shipperType == null
                || name == null || email == null || address == null) {
            throw new IllegalArgumentException("荷主の必須項目が欠けています");
        }
        this.id = id;
        this.shipperCode = shipperCode;
        this.shipperType = shipperType;
        this.name = name;
        this.email = email;
        this.phone = phone == null ? Phone.empty() : phone;
        this.address = address;
    }

    /** 個人荷主を新規登録する。 */
    public static Shipper registerIndividual(
            ShipperId id, ShipperCode code, ShipperName name,
            Email email, Phone phone, Address address) {
        return new Shipper(id, code, ShipperType.INDIVIDUAL, name, email, phone, address);
    }

    public boolean isCorporate() {
        return shipperType == ShipperType.CORPORATE;
    }

    public ShipperId id() {
        return id;
    }

    public ShipperCode shipperCode() {
        return shipperCode;
    }

    public ShipperType shipperType() {
        return shipperType;
    }

    public ShipperName name() {
        return name;
    }

    public Email email() {
        return email;
    }

    public Phone phone() {
        return phone;
    }

    public Address address() {
        return address;
    }
}

File: apps/cargo-tracker/packs/shipper/app/domain/shipper/domain/shipper.rb
# frozen_string_literal: true

module Shipper
  module Domain
    # 荷主集約ルート（PORO）。個人・法人の 2 種別を扱う。
    class Shipper
      attr_reader :code, :name, :email, :phone, :address, :shipper_type

      # 新規荷主を生成する。不変条件をここで担保する。
      def self.register(code:, name:, email:, shipper_type:, phone: nil, address: nil)
        raise ArgumentError, "荷主名は必須です" if name.nil? || name.to_s.strip.empty?
        raise ArgumentError, "メールアドレスは必須です" if email.nil? || email.to_s.strip.empty?

        new(code: code, name: name, email: email, phone: phone, address: address,
            shipper_type: shipper_type)
      end

      def initialize(code:, name:, email:, shipper_type:, phone: nil, address: nil)
        @code = code
        @name = name
        @email = email
        @phone = phone
        @address = address
        @shipper_type = shipper_type
      end

      def corporate?
        shipper_type.corporate?
      end
    end
  end
end
